urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+            } else {
+              bigpatch.diffs.getFirst().text = bigpatch.diffs.getFirst().text
+                  .substring(diff_text.length());
+            }
+          }
+        }
+        // Compute the head context for the next patch.
+        precontext = diff_text2(patch.diffs);
+        precontext = precontext.substring(Math.max(0, precontext.length()
+            - Patch_Margin));
+        // Append the end context for this patch.
+        if (diff_text1(bigpatch.diffs).length() > Patch_Margin) {
+          postcontext = diff_text1(bigpatch.diffs).substring(0, Patch_Margin);
+        } else {
+          postcontext = diff_text1(bigpatch.diffs);
+        }
+        if (postcontext.length() != 0) {
+          patch.length1 += postcontext.length();
+          patch.length2 += postcontext.length();
+          if (!patch.diffs.isEmpty()
+              && patch.diffs.getLast().operation == Operation.EQUAL) {
+            patch.diffs.getLast().text += postcontext;
+          } else {
+            patch.diffs.add(new Diff(Operation.EQUAL, postcontext));
+          }
+        }
+        if (!empty) {
+          pointer.add(patch);
+        }
+      }
+      bigpatch = pointer.hasNext() ? pointer.next() : null;
+    }
+  }
+
+  /**
+   * Take a list of patches and return a textual representation.
+   * @param patches List of Patch objects.
+   * @return Text representation of patches.
+   */
+  public String patch_toText(List<Patch> patches) {
+    StringBuilder text = new StringBuilder();
+    for (Patch aPatch : patches) {
+      text.append(aPatch);
+    }
+    return text.toString();
+  }
+
+  /**
+   * Parse a textual representation of patches and return a List of Patch
+   * objects.
+   * @param textline Text representation of patches.
+   * @return List of Patch objects.
+   * @throws IllegalArgumentException If invalid input.
+   */
+  public List<Patch> patch_fromText(String textline)
+      throws IllegalArgumentException {
+    List<Patch> patches = new LinkedList<Patch>();
+    if (textline.length() == 0) {
+      return patches;
+    }
+    List<String> textList = Arrays.asList(textline.split("\n"));
+    LinkedList<String> text = new LinkedList<String>(textList);
+    Patch patch;
+    Pattern patchHeader
+        = Pattern.compile("^@@ -(\\d+),?(\\d*) \\+(\\d+),?(\\d*) @@$");
+    Matcher m;
+    char sign;
+    String line;
+    while (!text.isEmpty()) {
+      m = patchHeader.matcher(text.getFirst());
+      if (!m.matches()) {
+        throw new IllegalArgumentException(
+            "Invalid patch string: " + text.getFirst());
+      }
+      patch = new Patch();
+      patches.add(patch);
+      patch.start1 = Integer.parseInt(m.group(1));
+      if (m.group(2).length() == 0) {
+        patch.start1--;
+        patch.length1 = 1;
+      } else if (m.group(2).equals("0")) {
+        patch.length1 = 0;
+      } else {
+        patch.start1--;
+        patch.length1 = Integer.parseInt(m.group(2));
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 2303:2311, 2314:2322.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.